### ISSUE TITLE
Explore including the type information into the group key to avoid schema collisions.

### ISSUE THREAD

**Issue by nathanielc** [influxdata/flux#439]
The idea is to avoid the mismatched type errors that we commonly encounter.

Update the SPEC with the proposed design.

**Comment by affo** [comment]
Related issues: #320 #321

**Comment by affo** [comment]
@nathanielc @wolffcm 
This is a summary of our conclusions from the last Slack call on type collision.  
I had a look at Guarrigue's paper on structural polymorphism (I think it would require a 2nd, 3rd, 4th, ... pass 😅 ).

We carefully analyzed the option of including tables schema in the group key and we drew some conclusions:
 - overall, the explanation in the SPEC seems pretty and concise (https://github.com/influxdata/flux/pull/556), but:
 - in implementing `Process` for some `Transformation`s, it is not always the case that you know in advance the schema of the output table (e.g. `pivot`). So, it becomes very impractical to get a builder from the `TableBuilderCache` by using the group key (that now should include the schema information). It would require multiple steps to compute and merge multiple tables to obtain the final result, thus making the implementation of that functions rather obscure and fragile.
 - Some simple aggregations behavior could become tricky, take, for example, `count`: suppose that you have a collision on `_value` that produces 2 distinct tables; suppose that you count; `_value` goes away and there is no type collision anymore. What should you do? Aggregate the results or not? The semantics of aggregation could become obscure.

We therefore decided to _drop the possibility of including the schema in the group key_ and keep the current implementation of erroring once a collision has been detected.
However, some problems are still there (#975) because there is not a unified implementation for regrouping that fails upon collision.

As @nathanielc suggested, we could do more. We could error only if the column that collided is accessed. The column would have a conjunction type (e.g. `int & float`) that is basically invalid. So, the column should have type `TInvalid`, and, once it is accessed, we should error and return the reason why it is invalid with a meaningful error message.
This is a great improvement: suppose that a user `drop`s `_value` after collision, the query could run smoothly.

The problem is also related to type inference.  
We already do type inference (https://github.com/influxdata/flux/blob/master/semantic/solve.go), but I find it difficult to infer (ah, the irony) at what extent. To my superficial understanding, we infer the types of variables from the AST (or, better, semantic representation), but we are not able to infer types of records in tables by analyzing the transformations in the graph of processing. Correct?


So, the inference mechanism, does not know what `map(fn: (r) => {lol: r._value + 1}, mergeKey: true)` means, and so it cannot infer what happens to the tables schema and what are the constraints on downstream transformations.
By introducing this feature, we could provide compile time errors for schema collision and more. However, we would still need, somehow, to get the schema from the source in order to error or not.

If this is a correct description of the problem, we have everything we need to proceed.  
We can isolate different implementation steps:
 1. We can fix every transformation that potentially causes collision (see https://github.com/influxdata/flux/pull/556#issuecomment-461140391);
 2. And/or provide a unified interface for regrouping and provide a proper error on collision;
 3. Use `TInvalid` to express conjunction types and error only when accessed;
 4. Add constraints information to transformations to perform type inference.

**Comment by nathanielc** [comment]
@chnn @pauldix and I just had another conversation about this.

Here are some notes:

* Use type information to disambiguate same named columns with different types
* Use an optional syntax to reference columns by their name and type, something like `r.usage_idle : float`

Combining these will allow for heterogeneous data to be consumed by clients.
The recent use case is to pagintate through records of heterogeneous data. See https://gist.github.com/chnn/62bdb41eaab5a8072800d0ae4d7783c7

In short if `group` can handle creating sparse tables (i.e. inject nulls) for heterogeneous data then the UI can use `group` + `limit` to paginate through the results.
